### DISCUSSION TITLE
How to ignore default locale prefix?

### DISCUSSION THREAD

**Discussion by sachithrrra** [amannn/next-intl#1166] [Q&A] [answered]
Is there a way to ignore the default locale prefix in the site?
I'd like to keep it like this for English: `https://example.com/about` instead of this: `https://example.com/en/about`

thanks!

middleware.ts
```
import createMiddleware from "next-intl/middleware";

export default createMiddleware({
  // A list of all locales that are supported
  locales: ["en", "de", "fr", "es", "it", "ja", "nl", "pl", "pt", "ru", "sv"],

  // Used when no locale matches
  defaultLocale: "en",
});

export const config = {
  // Match only internationalized pathnames
  matcher: ["/", "/(en|de|fr|es|it|ja|nl|pl|pt|ru|sv)/:path*"],
};
```

i18n.ts
```
import { getRequestConfig } from "next-intl/server";
import { notFound } from "next/navigation";

// Can be imported from a shared config
const locales = [
  "en",
  "de",
  "fr",
  "es",
  "it",
  "ja",
  "nl",
  "pl",
  "pt",
  "ru",
  "sv",
];

export default getRequestConfig(async ({ locale }) => {
  // Validate that the incoming `locale` parameter is valid
  if (!locales.includes(locale as any)) notFound();

  return {
    messages: (await import(`../messages/${locale}.json`)).default,
  };
});
```

config.ts
```
import { LocalePrefix, Pathnames } from "next-intl/routing";

export const defaultLocale = "en" as const;
export const locales = [
  "en",
  "de",
  "fr",
  "es",
  "it",
  "ja",
  "nl",
  "pl",
  "pt",
  "ru",
  "sv",
] as const;

export const pathnames: Pathnames<typeof locales> = {
  "/": "/",
};

export const localePrefix: LocalePrefix<typeof locales> = "always";

export const port = process.env.PORT || 3000;
export const host = process.env.VERCEL_URL
  ? `https://${process.env.VERCEL_URL}`
  : `http://localhost:${port}`;

```

**Comment by amannn** [comment]
Yes, you're looking for [`localePrefix: 'as-needed'`](https://next-intl-docs.vercel.app/docs/routing#locale-prefix-as-needed)

**Reply by stylehuan** [reply]
The localePrefix is already set to ‘as-needed’ but it’s not working

![image](https://github.com/amannn/next-intl/assets/6212099/1ab87625-edc3-4ae9-8ebb-f06f9ffa541f)

![image](https://github.com/amannn/next-intl/assets/6212099/a16435a9-e5d1-430b-81ca-d3940bbc83e2)

![image](https://github.com/amannn/next-intl/assets/6212099/1280d535-437d-420e-9cee-abcee1659178)

**Reply by amannn** [reply]
Two problems:

1. The `localePrefix` is not passed to the middleware
2. The matcher is off

 Please carefully review [the docs](https://next-intl-docs.vercel.app/docs/routing#locale-prefix-as-needed).

**Reply by stylehuan** [reply]
I’ve read the documentation many times, but I still don’t quite understand.

The localePrefix doesn’t seem to be used anywhere in the middleware.
How should I interpret “matcher is off”?

**Reply by shiny** [reply]
It's not working for me too. I'm new to next-intl, the experience of turing “as-needed” on was a bit bumpy.

**Comment by shiny** [comment, marked as answer]
Thanks to @amannn's guide, I've worked it out, the key is `middleware.ts`

```middleware.ts
import createMiddleware from "next-intl/middleware";

export default createMiddleware({
    locales: ["en", "de", "fr", "es", "it", "ja", "nl", "pl", "pt", "ru", "sv"],
    // 'as-needed' should pass to createMiddleware
    localePrefix: 'as-needed',
    defaultLocale: "en",
});

export const config = {
    // We should skip the internal routes
    matcher: ["/((?!api|static|.*\\..*|_next).*)", '/(zh-CN|en)/:path*']
};
```

**Comment by SamX23** [comment]
Thanks @shiny, 

Hi, @amannn let me know if my case is possible.

Let say, I want both "/" and "/en" working, so it should not automatically redirect to "/en" whenever I visit the "/" page.
Is this possible ?

**Reply by amannn** [reply]
You can configure the [`localePrefix`](https://next-intl-docs.vercel.app/docs/routing#locale-prefix) to either always use a prefix, or only when necessary (i.e. not for the default locale). Supporting both patterns at the same time is not supported, no.

Theoretically, you can configure the locale prefix of middleware dynamically by [composing the middleware](https://next-intl-docs.vercel.app/docs/routing/middleware#composing-other-middlewares), but I wouldn't recommend it.

**Reply by SamX23** [reply]
I see, thanks
